armory/adapt/pytorch.py

```diff
@@ -0,0 +1,1126 @@
+            self.delta.copy_(
+                self.delta_temp
+            )  # need to modify the 'delta' held in optimizer
+
+    def __call__(
+        self,
+        x_orig: torch.Tensor,
+        y_true=None,
+        y_target=None,
+        x_init=None,
+        iters=10,
+    ):
+        # TODO: verify x_orig in domain
+        iters = int(iters)
+        if iters < 0:
+            raise ValueError("iters must be nonnegative")
+        self.call_init(
+            early_stop=False,
+            best_x=None,
+            x_orig=x_orig,
+            y_true=y_true,
+            y_target=y_target,
+            x_init=x_init,
+            iters=iters,
+        )
+
+        if self.y_target is None:  # untargeted
+            self.targeted = False
+            if self.y_true is None:
+                self.y_target = self.model(self.x)
+            else:
+                self.y_target = self.y_true
+        else:
+            self.targeted = True
+
+        # TODO: improve placement of mask
+        #    For now, just use upper left
+        self.position = (0, 0, 0)
+        self.index = (slice(None),) + tuple(
+            slice(i, i + j) for (i, j) in zip(self.position, self.mask_size)
+        )
+        self.delta_orig = self.x_orig[self.index]
+        self.x_zero = self.x_orig.clone()
+        self.x_zero[self.index] = 0
+        self.delta = self.delta_orig.clone()
+        # check that x_orig is in domain
+        if x_init == "random":
+            self.delta = torch.rand(self.delta)
+        elif x_init is not None:
+            raise ValueError(f"x_init {x_init} must be 'random' or None")
+
+        self.optimizer = torch.optim.SGD(
+            [self.delta], lr=self.eps_step
+        )  # should be part of constructor
+
+        # modify delta (patch) instead of x directly
+        self.delta.requires_grad = True
+        for i in range(iters + 1):
+            self.i = i
+            self.optimizer.zero_grad()
+            self.gradient()
+            self.status()  # check for early stop, best value, bad values, etc.
+            if self.early_stop or self.i >= self.iters:
+                break
+
+            self.update()
+
+        # TODO: work on random placement of patch
+
+        return self.best_x
+
+
+def random_position(shape, mask_shape):
+    position = []
+    if isinstance(shape, torch.Tensor):
+        shape = shape.shape
+    if isinstance(mask_shape, torch.Tensor):
+        mask_shape = mask_shape.shape
+    if len(shape) > len(mask_shape) and shape[0] == 1:
+        # Assume batch
+        shape = shape[1:]
+    if len(shape) != len(mask_shape):
+        raise ValueError(f"shape {shape} not consistent with mask_shape {mask_shape}")
+    for i, j in zip(shape, mask_shape):
+        if j > i:
+            raise ValueError(f"mask_shape {mask_shape} exceeds shape {shape}")
+        position.append(np.random.randint(i - j + 1))
+    return tuple(position)
+
+
+class PGD_RandomPatch(PGD_Patch2):
```

Comment:
is it necessary to have an additional class for patch pgd _and_ random patch pgd, as opposed to this just being something set or defined in the former?